javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -104,9 +132,11 @@ private static int treeSize(int branchCount, int depthShift) {
         } else if (n >= length()) {
             return empty();
         } else {
-            final NodeModifier drop = Arrays::copyDrop;
-            final Object[] root = modifyLeaf(array, depthShift, offset + n, drop, drop);
-            return new BitMappedTrie<>(root, offset + n, length() - n, depthShift);
+            final int index = offset + n;
+            final Object[] root = arePointingToSameLeaf(0, n)
+                                  ? array
+                                  : modifyLeaf(array, depthShift, index, Arrays::copyDrop, Identity);
+            return new BitMappedTrie<T>(root, index, length() - n, depthShift).collapseHeight();
```

Comment:
I tried so many alternatives here (including the one you suggested), I didn't even know which one was committed in the end. This is why I needed an extra pair of eyes, thanks :).